File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        auto valid = [&](int mid) {
            int cnt = 0;
            for (auto pile: piles)
                cnt += (pile + mid - 1) / mid;
            return cnt <= h;
        };
        
        int lo = 1, hi = 1e9;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (valid(mid))
                hi = mid;
            else
                lo = mid + 1;
        }
        
        return lo;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def checker(self, piles, h, mid):
        num = 0
        for pile in piles:
            num += ceil(pile / mid)
        return num <= h
    
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low = 0
        high = max(piles) 
        
        while low + 1 < high:
            mid = low + (high - low) // 2
            if self.checker(piles, h , mid):
                high = mid
            else:
                low = mid
        return high        